service-runner.js

```diff
@@ -215,25 +256,82 @@ ServiceRunner.prototype._rollingRestart = function() {
     });
 };
 
+/*
+    This is a workaround for the following bug in node:
+    https://github.com/joyent/node/issues/9409
+
+    Shortly, the problem is that a worker is removed by master
+    if it sends 'disconnect' and then it's checked on 'exit'.
+    However, in some cases (for example an infinite CPU loop in worker,
+    'disconnect' may never happen, and master will crash receiving an 'exit'
+ */
+function fixCloseDisconnectListeners(worker) {
+    // Take the exit listener node's cluster have set.
+    var listeners = worker.process.listeners('exit')[0];
+    var exit = listeners[Object.keys(listeners)[0]];
+
+    // Take the disconnect litener node's cluster have set.
+    listeners = worker.process.listeners('disconnect')[0];
+    var disconnect = listeners[Object.keys(listeners)[0]];
+
+    // Now replace the exit listener to make sure 'disconnect' was called
+    worker.process.removeListener('exit', exit);
+    worker.process.once('exit', function(exitCode, signalCode) {
+        if (worker.state != 'disconnected') {
+            disconnect();
+        }
+        exit(exitCode, signalCode);
+    });
+}
+
+/**
+ * Saves the timestamp of a worker heartbeat
+ * @private
+ */
+ServiceRunner.prototype._saveBeat = function(worker) {
+    var self = this;
+    var currentVal = self.workerHeartbeatTime[worker.process.pid];
+    self.workerHeartbeatTime[worker.process.pid] = {
```

Comment:
What's the point of updating the time if the worker's `killed` flag is set to true? IMHO, we should avoid allocating a new object in that case, i.e.

```js
if (currentVal && currentVal.killed) {
    return;
}
```